libs/payments/cart/src/lib/checkout.service.ts

```diff
@@ -2,43 +2,102 @@
   ) {}
 
   async prePaySteps(cart: ResultCart) {
-    // TODO:
-    // - If uid not present, create user stub account and update cart
-    // - Check if customer exists, create via stripe manager and update cart
-    // - Make sure shipping address of customer is updated with cart taxAddress
-    // - Check if customer is eligible for product in cart
-    // - Check if amount to be charged matches amount in cart
-    // - Cancel incomplete subscriptions to price user is subscribing to
-
-    const { stripeCustomerId } = cart;
-    if (!stripeCustomerId) {
-      // TODO: this is a placeholder and won't be necessary once the above todo items are done
-      throw new Error("Stripe customer doesn't exist");
+    let customer;
+
+    // if stripeCustomerId does not exist:
+    // create stub via StripeClient and update cart
+    if (!cart.uid || !cart.stripeCustomerId) {
+      const customerStub = await this.stripeClient.customersCreate();
+
+      await this.cartManager.updateFreshCart(cart.id, cart.version, {
+        stripeCustomerId: customerStub.id,
+      });
+
+      customer = await this.stripeManager.fetchActiveCustomer(customerStub.id);
+    } else {
+      customer = await this.stripeManager.fetchActiveCustomer(
+        cart.stripeCustomerId
+      );
     }
-    const customer = await this.stripeManager.fetchActiveCustomer(
-      stripeCustomerId
+
+    // update customer shipping address with cart taxAddress
+    const taxAddress = cart.taxAddress as any as TaxAddress;
+
+    await this.stripeManager.updateCustomer(customer.id, {
+      address: {
+        country: taxAddress.countryCode,
+        postal_code: taxAddress.postalCode,
+      },
+    });
+
+    // validate customer is eligible for product via eligibility service
+    // throws cart eligibility mismatch error if no match found
+    const eligibility = await this.eligibilityService.checkEligibility(
+      cart.interval,
+      cart.offeringConfigId,
+      customer.id
     );
 
+    const cartEligibilityStatus = handleEligibilityStatusMap[eligibility];
+
+    if (cartEligibilityStatus !== cart.eligibilityStatus) {
+      throw new CartEligibilityMismatchError(cart.id);
+    }
+
+    // re-validate total amount against upcoming invoice via StripeClient
+    // throws cart total mismatch error if no match found
+    const upcomingInvoice = await this.stripeClient.invoicesRetrieveUpcoming();
+
+    if (upcomingInvoice.total !== cart.amount) {
+      throw new CartTotalMismatchError(cart.id);
+    }
+
+    // check if customer already has subscription to price and cancel if they do
+    const subscriptions = await this.stripeManager.getSubscriptions(
+      customer.id
+    );
+
+    const subscription = subscriptions.find(
+      (subscription: StripeSubscription) =>
+        subscription.id === cart.offeringConfigId
+    );
+
+    if (subscription) {
+      await this.stripeManager.cancelIncompleteSubscriptionsToPrice(
+        customer.id,
+        cart.offeringConfigId
```

Comment:
`cancelIncompleteSubscriptionsToPrice` seems to require a customerId, which is being provided correctly, and priceId, which unfortunately is not.

See comments above regarding `offeringConfigId`. Also see comments above on how to retrieve the priceId using offeringConfigId and Interval.